### PR TITLE
feat(ui5-app-uinquirer): prohibit non cap fiori generation into a cap project folder

### PR DESCRIPTION
**Pull request by korotkovao**
https://github.com/SAP/open-ux-tools/issues/2427

Alert and prevent from generating a Fiori application into a CAP project folder unless I have chosen the Local CAP Project as my data source.

### CODE REVIEW

File: packages/ui5-application-inquirer/src/prompts/validators.ts [hunk 1]
```diff
@@ -32,7 +32,14 @@ export function validateAppName(appName: string, targetDir: string): boolean | s
 export async function validateFioriAppProjectFolder(targetDir: string): Promise<string | boolean> {
     const appRoot = await findRootsForPath(targetDir);
     if (appRoot) {
-        return t('validators.folderContainsFioriApp', { path: appRoot.appRoot });
+        return t('validators.folderContainsFioriApp', { path: appRoot?.appRoot });
+    }
+    // Check targetDir is in CAP project sub directories
+    const isCapPath = await findCapProjectRoot(targetDir);
```
Comment (IainSAP) [inline packages/ui5-application-inquirer/src/prompts/validators.ts:38]: Should we instead create a new function in project-access to determine if the path is in a CAP project. e.g. `isCapPath(targetDir)`

File: packages/ui5-application-inquirer/src/prompts/validators.ts [hunk 2]
```diff
@@ -32,7 +32,14 @@ export function validateAppName(appName: string, targetDir: string): boolean | s
 export async function validateFioriAppProjectFolder(targetDir: string): Promise<string | boolean> {
     const appRoot = await findRootsForPath(targetDir);
     if (appRoot) {
-        return t('validators.folderContainsFioriApp', { path: appRoot.appRoot });
+        return t('validators.folderContainsFioriApp', { path: appRoot?.appRoot });
+    }
+    // Check targetDir is in CAP project sub directories && check if targetDir is a CAP project root directory
+    const isCapPath = (await findCapProjectRoot(targetDir)) && (await isCapProject(targetDir));
+    // Check targetDir is a CAP project root directory
+    //const CapType = await getCapProjectType(targetDir);
```
Comment (docirl) [inline packages/ui5-application-inquirer/src/prompts/validators.ts:40]: Remove commented line?

File: packages/ui5-application-inquirer/src/prompts/validators.ts [hunk 3]
```diff
@@ -1,7 +1,7 @@
 import { validateModuleName } from '@sap-ux/project-input-validator';
 import { appPathExists } from './prompt-helpers';
 import { t } from '../i18n';
-import { findRootsForPath } from '@sap-ux/project-access';
+import { findCapProjectRoot, findRootsForPath, getCapProjectType, isCapProject } from '@sap-ux/project-access';
```
Comment (docirl) [inline packages/ui5-application-inquirer/src/prompts/validators.ts:4]: `getCapProjectType` import no longer needed (code smell)

File: packages/ui5-application-inquirer/src/prompts/validators.ts (hunk 2 above)
Comment (korotkovao) [reply packages/ui5-application-inquirer/src/prompts/validators.ts]: Thank you, i have removed it :)

File: packages/ui5-application-inquirer/src/prompts/validators.ts (hunk 3 above)
Comment (korotkovao) [reply packages/ui5-application-inquirer/src/prompts/validators.ts]: Thank you, i have removed it :)

File: packages/ui5-application-inquirer/src/prompts/validators.ts (hunk 1 above)
Comment (korotkovao) [reply packages/ui5-application-inquirer/src/prompts/validators.ts]: Thank you, I have updated the code to use `isCapProject()`

File: packages/project-access/src/project/search.ts [hunk 4]
```diff
@@ -499,3 +499,27 @@ export async function findCapProjects(options: {
...
 }
+
+/**
+ * Returns true if the specified target path contains a CAP project.
+ * Checks sub and root directories.
+ *
+ * @param targetDir the target directory path.
+ * @returns true if CAP Project is in the directory or false if not.
+ */
+export async function checkPathForCapApp(targetDir: string): Promise<string | boolean> {
```
Comment (cianmSAP) [inline packages/project-access/src/project/search.ts:510]: We could updated (simplified) to be 

```
export async function checkPathForCapApp(targetDir: string): Promise<boolean> {
    // Check if targetDir is a CAP project
    const capType = await getCapProjectType(targetDir);

    // If not a CAP project root, check if the CAP root is in the subdirectories
    return !!capType || !!(await findCapProjectRoot(targetDir));
}
```

File: packages/project-access/test/project/cap.test.ts [hunk 5]
```diff
@@ -88,6 +89,14 @@ describe('Test isCapJavaProject()', () => {
         ).toBeFalsy();
     });
 });
+describe('Test checkPathForCapApp()', () => {
```
Comment (cianmSAP) [inline packages/project-access/test/project/cap.test.ts:92]: You could add a couple more tests here, one where cap project is the root, where the cap root is in the sub dirs, and then a negative test

File: packages/ui5-application-inquirer/src/prompts/validators.ts [hunk 6]
```diff
@@ -32,7 +32,11 @@ export function validateAppName(appName: string, targetDir: string): boolean | s
 export async function validateFioriAppProjectFolder(targetDir: string): Promise<string | boolean> {
     const appRoot = await findRootsForPath(targetDir);
     if (appRoot) {
-        return t('validators.folderContainsFioriApp', { path: appRoot.appRoot });
+        return t('validators.folderContainsFioriApp', { path: appRoot?.appRoot });
```
Comment (cianmSAP) [inline packages/ui5-application-inquirer/src/prompts/validators.ts:35]: I don't think there's any need for the `?` in `{ path: appRoot?.appRoot }` becuase you already check if it's defined

**Review by cianmSAP — CHANGES_REQUESTED** [review]
Just a couple of small requests

File: packages/project-access/src/project/search.ts (hunk 4 above)
Comment (cianmSAP) [reply packages/project-access/src/project/search.ts:510]: The comments can probably be removed, the JSDoc can explain the process

Comment (korotkovao) [reply packages/project-access/src/project/search.ts:510]: Thank you for the review.
I agree, however, multiple colleagues were asking about the need for the two functions to do a similar thing so I might leave the comments if that is okay?

File: packages/project-access/test/project/cap.test.ts (hunk 5 above)
Comment (korotkovao) [reply packages/project-access/test/project/cap.test.ts:92]: Thank you for the review. Added some.

File: packages/ui5-application-inquirer/src/prompts/validators.ts (hunk 6 above)
Comment (korotkovao) [reply packages/ui5-application-inquirer/src/prompts/validators.ts]: Thank you for the review. Removed as suggested.

File: packages/project-access/src/project/search.ts (hunk 4 above)
Comment (cianmSAP) [reply packages/project-access/src/project/search.ts:510]: Ok, sure

File: packages/project-access/src/project/search.ts [hunk 7]
```diff
@@ -499,3 +499,18 @@ export async function findCapProjects(options: {
...
```
Comment (heimwege) [inline packages/project-access/src/project/search.ts:510]: nit-pick: I would prefer to name this function something like `isPathForCapApp`. There's not a real check being done and you also named the property in `validators.ts` accordingly `isCapPath`. So why not the whole function?

```ts
const isCapPath = await checkPathForCapApp(targetDir);
```
looks a bit odd from my point of view

File: packages/ui5-application-inquirer/src/prompts/validators.ts [hunk 8]
```diff
@@ -33,6 +33,10 @@ export async function validateFioriAppProjectFolder(targetDir: string): Promise<
     const appRoot = await findRootsForPath(targetDir);
     if (appRoot) {
         return t('validators.folderContainsFioriApp', { path: appRoot.appRoot });
+    } // Check targetDir is in CAP project sub directories && check if targetDir is a CAP project root directory
```
Comment (heimwege) [inline packages/ui5-application-inquirer/src/prompts/validators.ts:36]: This comment is superfluous and should be covered by the JSdoc of `checkPathForCapApp`

File: packages/ui5-application-inquirer/src/prompts/validators.ts [hunk 9]
```diff
@@ -33,6 +33,10 @@ export async function validateFioriAppProjectFolder(targetDir: string): Promise<
...
+    const isCapPath = await checkPathForCapApp(targetDir);
+    if (isCapPath) {
```
Comment (heimwege) [inline packages/ui5-application-inquirer/src/prompts/validators.ts:38]: nit-pick: Is there an actual need for `const isCapPath` or can this be simplified to 

```ts
if (await checkPathForCapApp(targetDir)) {
```

File: packages/project-access/src/project/search.ts [hunk 10]
```diff
@@ -499,3 +499,18 @@ export async function findCapProjects(options: {
...
+    // Check if targetDir is a CAP project
+    const capType = await getCapProjectType(targetDir);
+
+    // If not a CAP project root, check if the CAP root is in the subdirectories
+    return !!capType || !!(await findCapProjectRoot(targetDir));
```
Comment (heimwege) [inline packages/project-access/src/project/search.ts:515]: nit-pick: check can be simplified to

```ts
return !!(await getCapProjectType(targetDir)) || !!(await findCapProjectRoot(targetDir));
```

**Review by heimwege — COMMENTED** [review]
Only nit-picks. Sorry for that 🙈

File: packages/project-access/src/project/search.ts (hunk 7 above)
Comment (korotkovao) [reply packages/project-access/src/project/search.ts]: Thank you, changed as requested.

File: packages/ui5-application-inquirer/src/prompts/validators.ts (hunk 8 above)
Comment (korotkovao) [reply packages/ui5-application-inquirer/src/prompts/validators.ts]: Thank you, changed as requested.

File: packages/ui5-application-inquirer/src/prompts/validators.ts (hunk 9 above)
Comment (korotkovao) [reply packages/ui5-application-inquirer/src/prompts/validators.ts]: Thank you, changed as requested.

File: packages/project-access/src/project/search.ts (hunk 10 above)
Comment (korotkovao) [reply packages/project-access/src/project/search.ts]: Thank you, changed as requested.

**Review by devinea — APPROVED** [review]
Code changes are clear and covered by tests
changeset ✅

**Review by cianmSAP — APPROVED** [review]
Thanks @korotkovao 

Comments addressed
Code looks good
Excellent test coverage
Changeset
